### DISCUSSION TITLE
I am a beginner to Ebiten

### DISCUSSION THREAD

**Discussion by DAT4** [hajimehoshi/ebiten#1525] [Q&A] [answered]
Hi, I am a beginner to Ebiten and game development in general. And I have a lot of questions! 

I found your cheet sheet and it is really awesome, and really helpful. It got me started very fast.

I made a simple game now using go-websocket for multiplaying, by sharing commands in text.
I don't know if that is a good practice, but so far it is working and without any noticeable delay. 

I don't know if GEO.Translate is the best function to use for the movement because there is a chance that the players actual coordinates will be wrong if some messages are lost ex. if the player disconnects and connects again then the other players will be in another unknown position.
Would it be better just to send the actual coordinates of the player over the websocket, and use them and in that case. Is there a function I can use to put the player in that spot instead of translating all the time?

Also I had some problems when I was trying to compile to wasm. Most of the time I got errors in the browser, and then when I got it to work then it didn't register me pressing the buttons for movement.

Most of all I was wondering if there is a way to present text and to get userinput in a nice way.
For example I would like to let the user login with his own username and password before he can start the game. 
Also there is some choices he has to make to be able to start the game. So for example pressing a button which character he wants to be.
What would be the best way to implement interaction like this (userinput and menus/buttons)? 

My game is here:
https://github.com/DAT4/go-game

**Comment by hajimehoshi** [comment, marked as answer]
Hi, thank you for using Ebiten!

> Would it be better just to send the actual coordinates of the player over the websocket, and use them and in that case. Is there a function I can use to put the player in that spot instead of translating all the time?

This sounds like a general game system design question, rather than a question about Ebiten. I don't have a good answer but I'd make the server have the master data and each client render them as the easiest way.

> Also I had some problems when I was trying to compile to wasm. Most of the time I got errors in the browser, and then when I got it to work then it didn't register me pressing the buttons for movement.

What kind of error did you get?

> Most of all I was wondering if there is a way to present text and to get userinput in a nice way.

What about [`ebiten.InputChars`](https://pkg.go.dev/github.com/hajimehoshi/ebiten/v2#InputChars)?

**Reply by hajimehoshi** [reply]
To present a text, use [`text`](https://pkg.go.dev/github.com/hajimehoshi/ebiten/v2/text) package. See [`font`](https://ebiten.org/examples/font.html) example for its usage.

**Reply by DAT4** [reply]
Thank you for your reply. 

1. Should I use [`GeoM.apply(x,y)`](https://pkg.go.dev/github.com/hajimehoshi/ebiten/v2#GeoM.Apply) to set the x and y coordinates? or something else?
2. I think I got some CORS errors about the wasm, and then I disabled CORS in the browser and then I got some other errors like this: 
`unsupported MIME type 'application/octet-stream' expected 'application/wasm'` I think it has somthing to do with me getting userinput from the commandline as well, but then I tried to remove that. Yet I still got the same error

Thank you. I am already using `ebiten.InputChars()` for chat. I was using the example from the site.
I was wondering how to present text in a bigger size than I can with the `ebitenutil.DebugPrint()` which I am using now.

:)

**Reply by DAT4** [reply]
Wow thank you I will look at the `text` package an `font` example

**Reply by hajimehoshi** [reply]
> Should I use GeoM.apply(x,y) to set the x and y coordinates? or something else?

Yes!

> I think I got some CORS errors about the wasm, and then I disabled CORS in the browser and then I got some other errors like this: unsupported MIME type 'application/octet-stream' expected 'application/wasm' I think it has somthing to do with me getting userinput from the commandline as well, but then I tried to remove that. Yet I still got the same error

Please take a look at your server's configuration. The message indicates Wasm must be served with `application/wasm`

> I was wondering how to present text in a bigger size than I can with the ebitenutil.DebugPrint() which I am using now.

Already answered above :)

**Reply by DAT4** [reply]
Thank you. And thanks again I really love Ebiten already. Yet I think the internet needs more tutorials about Ebiten. Maybe Ill make one when I learn it myself. :)

**Reply by hajimehoshi** [reply]
Yeah, writing more documentation is one of tasks we should work on. Thank you for trying Ebiten again!

**Reply by DAT4** [reply]
Sorry maybe this is a stupid question. But howcome the GeoM type doesnt just have a function which can update the x and y coordinates directly.? 

I see that [`GeoM.Translate()`](https://github.com/hajimehoshi/ebiten/blob/c536e34e56223c5680e1323198d0634ae31db686/geom.go#L119) is just making a relative change.
```go
func (g *GeoM) Translate(tx, ty float64) {
	g.tx += tx
	g.ty += ty
}
```
Why not giving the user more control and making a similar function

```go
func (g *GeoM) Change(tx, ty float64) {
	g.tx = tx
	g.ty = ty
}
```

That way I can have my server handle the logic and then send the coordinates to all the clients.

You mentioned that I should just use [`GeoM.Apply()`](https://github.com/hajimehoshi/ebiten/blob/c536e34e56223c5680e1323198d0634ae31db686/geom.go#L55) 

I don't understand what I can use that function for because it does not change the object. Do you mean that I should use it in a combination with Translate? and in that case. Isnt it just the same at updating the `g.tx` and `g.ty` directly?

Sorry for spamming you. Maybe I should just learn more about matrices before I start game development. 

Best regards 
Martin

**Reply by AWachtendorf** [reply]
You should extract these to variables and just point to the fields. I usually create structs with variables that can be updated.

So the GeoM always is translated an these coordinates, and in the update function of ebiten you can then change these variables via a pointer. Does that help, or should I better provide a code example?

**Reply by DAT4** [reply]
Thank you. I don't quite follow how you can update the x and y coordinates in the GeoM because they are private. If you wouldn't mind I would very much like an example. Thank you :)

**Reply by DAT4** [reply]
Right now I am doing the same as in [the filter example](https://ebiten.org/examples/filter.html) Where a new instance of `DrawImageOptions{}` are created in the Draw function. It works, and I don't feel anything slowing down but I think it would be possible to optimize if I could just have one instance of the struct and then modify the numbers. 

Maybe I am wrong :)

**Reply by AWachtendorf** [reply]
Okay, to most of the examples create some objects that are then called globally in the Draw and Update functions. Whihc is fine. But if you want to include more structs to work with I usually go the way you see below. I hope you can read it, as  I cant get the bloody syntax highlighting to work. (Any tip for that? Might just be something I do wrong in github... )

// Ship could be a spaceship.
type Ship struct {
	shipimg          *ebiten.Image
	shipImageOptions *ebiten.DrawImageOptions
	posX, posY       float64
}

// DrawShipOnScreen is for translating our Ship Image.
func (s *Ship) DrawShipOnScreen(screen *ebiten.Image) {
	s.shipImageOptions.GeoM.Reset() // reset matrix so it does not get updated in endless loop 
	s.shipImageOptions.GeoM.Translate(-s.width/2, -s.height/2) // translate at center of image
	s.shipImageOptions.GeoM.Translate(s.posX, s.posY) //image is always translated to those koordinates
	screen.DrawImage(s.shipImg, s.shipImageOptions) // Draw the image
}

// Update listens to pressed keys and changes the posX,posY values.
// As those are pointed to the *Ship struct our drawn image is moved
func (s *Ship) Update() error {
	if ebiten.IsKeyPressed(ebiten.KeyW) {
		s.posY -= 0.5
	}

	if ebiten.IsKeyPressed(ebiten.KeyS) {
		s.posY += 0.5
	}
}


// The Game struct can then include a *Ship
type Game struct {
	ship *Ship
}

// Update is called constantly, this is the place where I let
// the game listen to changes 
func (g *Game) Update() error {
	g.ship.Update()
}

// Draw now calls the Draw function in our ship
func (g *Game) Draw(screen *ebiten.Image) {
	g.ship.DrawShipOnScreen(screen)
}

**Reply by hajimehoshi** [reply]
Just a quick reply:

> You mentioned that I should just use GeoM.Apply()

Oops sorry, I misread that. You should use `GeoM.Translate`

**Reply by hajimehoshi** [reply]
So I think you want to put an image on an absolute position, right? I think you are trying to reuse `DrawImageOptions` objects, but in this case you don't have to. You can create `DrawImageOptions` on the fly. Even if you want to reuse them for performance reasons, you can call `GeoM.Reset()` so that all values become 0.

**Reply by DAT4** [reply]
Thank you. :)
